scrapy_zyte_api/handler.py

```diff
@@ -86,6 +114,149 @@ def _prepare_api_params(self, request: Request) -> Optional[dict]:
+                )
+            elif request.headers:
+                headers = []
+                for k, v in request.headers.items():
+                    if not v:
+                        continue
+                    v = b",".join(v).decode()
+                    lowercase_k = k.strip().lower()
+                    if lowercase_k in self._unsupported_headers:
+                        if lowercase_k != b"user-agent" or v != DEFAULT_USER_AGENT:
+                            logger.warning(
+                                f"Request {request} defines header {k}, which "
+                                f"cannot be mapped into the Zyte Data API "
+                                f"customHttpRequestHeaders parameter."
+                            )
+                        continue
+                    k = k.decode()
+                    headers.append({"name": k, "value": v})
+                if headers:
+                    api_params["customHttpRequestHeaders"] = headers
+        if not response_body or any(
+            api_params.get(k) for k in ("browserHtml", "screenshot")
+        ):
+            headers = api_params.get("requestHeaders")
+            if headers is not None:
+                logger.warning(
+                    f"Request {request} defines the Zyte Data API "
+                    f"requestHeaders parameter, overriding Request.headers. "
+                    f"Use Request.headers instead."
+                )
+            elif request.headers:
+                request_headers = {}
+                for k, v in request.headers.items():
+                    if not v:
+                        continue
+                    v = b",".join(v).decode()
+                    lowercase_k = k.strip().lower()
+                    key = self._browser_headers.get(lowercase_k)
+                    if key is not None:
+                        request_headers[key] = v
+                    elif not (
+                        (
+                            lowercase_k == b"accept"
+                            and v == DEFAULT_REQUEST_HEADERS["Accept"]
+                        )
+                        or (
+                            lowercase_k == b"accept-language"
+                            and v == DEFAULT_REQUEST_HEADERS["Accept-Language"]
+                        )
+                        or (lowercase_k == b"user-agent" and v == DEFAULT_USER_AGENT)
+                    ):
+                        logger.warning(
+                            f"Request {request} defines header {k}, which "
+                            f"cannot be mapped into the Zyte Data API "
+                            f"requestHeaders parameter."
+                        )
+                if request_headers:
+                    api_params["requestHeaders"] = request_headers
+
+    def _map_params(self, api_params, request):
+        if not any(
+            api_params.get(k) for k in ("httpResponseBody", "browserHtml", "screenshot")
```

Comment:
this is the piece I'm least certain about, from the API point of view. Let's say automatic mapping is enabled. User sends the following request:

```py
yield Request(url, meta={"zyte_api": {"screenshot": True}})
```

Would you expect to get a screenshot in addition to the usual response, or would you expect to get screenshot only? If that's screenshot only, should response body be set to the screenshot content?

What happens if we add data extraction later, e.g. 

```py
yield Request(url, meta={"zyte_api": {"product": True}})
```

If we follow the same approach, would it mean that response is going to be empty, and only product data would be returned?

If we're in a "Zyte API is my downloader" mindset, it's arguably expected that `response` in a callback still works as usual, but there is an additional data in `response` attributes (currently in `response.raw_api_response`, but we can improve on that).

This may lead to some inefficient code though, when users only want a screenshot or data, but we automatically request httpResponseBody as well.

Sorry for rambling, I'm still processing it :) 

I have a vague idea that making automatic mode completely separate from manual mode could help - i.e. instead of using `zyte_api` meta key for both modes, use separate meta keys for them (e.g. `zyte_api` and `zyte_api_raw`), as semantics of the parameters there are different. But it needs more thought.

Reply:
> ```python
> yield Request(url, meta={"zyte_api": {"screenshot": True}})
> ```
> 
> Would you expect to get a screenshot in addition to the usual response, or would you expect to get screenshot only?

I think we can choose what we think will be the most commonly desired scenario, and make that the default, as long as we document it. Similar to what we do with response headers.

For example, let’s say we believe that most commonly when users set `"screenshot": True` they will also want `browserHtml`. We can set `browserHtml` to `True` in those scenarios, but still allow to override that setting `browserHtml` to `False`.

I think that behavior would be acceptable as long as we document it.

However, since to get `browserHtml` users need, with the current implementation, to explicitly request it, I think in this specific case it may be best for `screenshot` to get only the screenshot, and not browser HTML. And, as you mention later on:

> This may lead to some inefficient code though, when users only want a screenshot or data, but we automatically request httpResponseBody as well.

---

> If that's screenshot only, should response body be set to the screenshot content?
>
> What happens if we add data extraction later, e.g.
> 
> ```python
> yield Request(url, meta={"zyte_api": {"product": True}})
> ```
> 
> If we follow the same approach, would it mean that response is going to be empty, and only product data would be returned?

I think this question is valid regardless of whether or not automatic mapping is enabled. So I don’t think we need an answer as part of this pull request.

I have no strong opinion on what the right behavior should be.

I think it would be weird for response to be anything other than an HTTP response, so if you send a request for e.g. only an automatically-extracted product, I don’t think it needs to become the response, I think an empty response with `.zyte_api_raw['product']` or directly `.product` may be best. But it could become the response as JSON, I guess.

For a future where you can combine browserHtml with httpResponseBody and we want users to be able to control which one becomes the response (which I am not sure is needed, honestly), or in case we want to allow screenshot to become the response, we could have a default implementation like “browser, else http, else screenshot” and expose a request meta key and setting to change that per-request or globally (e.g. `zyte_api_body_source="screenshot"`).

> I have a vague idea that making automatic mode completely separate from manual mode could help - i.e. instead of using `zyte_api` meta key for both modes, use separate meta keys for them (e.g. `zyte_api` and `zyte_api_raw`), as semantics of these parameters are different. But it needs more thought.

It may be a good approach. It is shorter than having a separate `automap` meta key. I do wonder how well it would play with other settings like `DEFAULT_PARAMS`.

Shall I extend tests (and implementation) to explore the scenarios with this approach?

Reply:
By the way, if the future-proofing settings are not too much, we could consider adding one that determines what keywords are considered content-defining, so that `httpResponseBody ` is not added if they are added. Thought that may be going too far, and it opens up other questions: Do we need to have separate settings to define which content keywords are compatible with browserHtml and which are not? Do we need to do the same for httpResponseBody? Do we need to allow configuring content keywords incompatible with both? Do we need to allow defining additional content keywords that are incompatible with each other?

Reply:
It looks like there are 2-3 significantly different use cases:

1. Use Zyte API as a replacement for Scrapy downloader, essentially as a proxy.
2. Everything else, specialized use cases: screenshots, data extraction.
3. A combined use case, where you want both direct download results and some extra data, e.g. screenshot or extracted data.

browserHtml is somewhere in between (1) and (2), probably closer to the "proxy" side.

The best API for these use cases could be different. 

For example, for (1) you'd like all requests to be passed through Zyte API implicitly. You'd also like to support various scrapy.Request parameters, and behave in a transparent way, in general. Sometimes you may need to e.g. replace httpResponseBody with browserHtml, set actions, and maybe tweak geolocation. To access the data, you'd want to use response.body/text, or run response.css/xpath selectors.

For (2) you want only sepecific requests to be handled, it should be opt-in. Likely you'd want to always specify request parameters, such as data type to extract, or screenshot size. It's not clear how useful is it to support Request parameters other than URL; it may be nice in some cases, but there's a lot of caveats. It doesn't make much sense to put extracted data to response body, and I think the same applies for screenshots.

For (3) you'd like scrapy.Request and response parameters to be mapped transparently, like in (1). In addition to that, you'd need to specify request parameters, and have this mode opt-in, like in (2).

What do you think about the following proposal?

1. Add an option like `ZYTE_API_DOWNLOAD=True`. This enables Mode (1) - automatic mapping of scrapy.Request, etc. For the user it looks like everything works as before, but Zyte API is now used to download data.
2. Add a request.meta key (something like `zyte_api_extra`), which allows to add more things to the API request: set geolocation, request screenshot. When something is present in this meta key, it's an extra, the default httpResponseBody is still present. This enables configuration for Mode (1), and it also enables Mode (3).
3. When `zyte_api` meta key is present, disable Mode 1, if it's enabled and use parameters in`zyte_api` as-is. This is essentially Mode (2). It should be backwards compatible to what we have in the main branch.
4. When `browserHtml` is present in `zyte_api_extra`, switch to "browser mode" - don't request httpResponseBody, use browserHtml as response body. I.e. returning both httpResponseBody and browserHtml in the same request is not supported with Mode (1); users can do it using Mode (2), with `zyte_api` meta keys.

Reply:
> Add an option like `ZYTE_API_DOWNLOAD=True`

Naming-wise, don’t you prefer `ZYTE_API_AUTOMAP` for this, or something in those lines (e.g. TRANSPARENT_MODE, PROXY_MODE)? To me `DOWNLOAD` feels generic enough that it could be confusing, since you can “download” content with `DOWNLOAD=False`, i.e. someone that still does not know what it does would wonder how `ENABLED` and `DOWNLOAD` are different.

> browserHtml handling in the proposal above is a bit weak.

I am not sure what you mean here, it sounds OK to me.

---

Besides naming, I am OK with this approach of split meta keys, I’ll go for it.

Reply:
> Naming-wise, don’t you prefer ZYTE_API_AUTOMAP for this, or something in those lines (e.g. TRANSPARENT_MODE, PROXY_MODE)? To me DOWNLOAD feels generic enough that it could be confusing, since you can “download” content with DOWNLOAD=False, i.e. someone that still does not know what it does would wonder how ENABLED and DOWNLOAD are different.

Naming.. :) My thinking was: we should have a name which communicates that "turn this option ON - now all your downloads are handled by Zyte API". `..PROXY_MODE` was my first thought, but it might get a bit confusing, as Zyte API is not a proxy now (and it's possible it'd get a proxy interface later).  I'm fine with `ZYTE_API_TRANSPARENT_MODE` or some other name. "AUTOMAP" sounds more low-level. 

By the way, in the proposal above, it's a single option, which enables automatic mapping and enables Zyte API on all requests. But we shuld have a way to disable it per-request, I haven't thought about it.

> I am not sure what you mean here, it sounds OK to me.

I'm not fond of `browserHtml` parameter behaving very different from other parameters in `zyte_api_extra`. For others, you can just add them to API request, but `browserHtml` changes the behavior completely. But it might be ok.

Reply:
> 1. Add an option like `ZYTE_API_DOWNLOAD=True`. This enables Mode (1) - automatic mapping of scrapy.Request, etc. For the user it looks like everything works as before, but Zyte API is now used to download data.

The way you describe it, it sounds to me like `ZYTE_API_TRANSPARENT_MODE` would not only activate automated mapping of request fields, but also implicitly set `ZYTE_API_ON_ALL_REQUESTS=True`.

At first I thought you wanted to merge both settings, but I think a setting like `ZYTE_API_ON_ALL_REQUESTS` still has its place, specially in combination with `ZYTE_API_DEFAULT_PARAMS`, so I am not sure what the best way forward is here. `settings.get("ZYTE_API_ON_ALL_REQUESTS", settings.get("ZYTE_API_TRANSPARENT_MODE", False))`?

I also wonder, if users want automated mapping for specific requests, what would the `meta` look like? `"zyte_api_transparent_mode": True, "zyte_api_extra": {"browserHtml": True}`? I think it would be nice if you did not need 2 keys for this, but having `"zyte_api_extra"` with a value other than `False` enable transparent mode feels wrong naming-wise. I wonder if `"zyte_api_transparent_mode": {"browserHtml": True}` would be OK.

Reply:
> The way you describe it, it sounds to me like ZYTE_API_TRANSPARENT_MODE would not only activate automated mapping of request fields, but also implicitly set ZYTE_API_ON_ALL_REQUESTS=True

Yes, kind-of. I was thinking there is no really a need for `ZYTE_API_ON_ALL_REQUESTS=True` if user uses "explicit" mode (data extraction, screenshots) - in these cases you'd likely want to enable things manually. If you don't do it, you can get screenshots of sitemaps, or attempted data extraction from robots.txt or media files. 

If `ZYTE_API_TRANSPARENT_MODE=True`, then I don't see a use case for disabling `ZYTE_API_ON_ALL_REQUESTS`. So, `ZYTE_API_ON_ALL_REQUESTS` doesn't look vary useful if we have transparent mode.

With that, it seems the main question is how ZYTE_API_DEFAULT_PARAMS behaves. We have 2 options:

1. ZYTE_API_DEFAULT_PARAMS are applied both for "transparent" and "explicit" requests. For example, one may want to set geolocation for all requests.
2. ZYTE_API_DEFAULT_PARAMS is only applied for "explicit" `meta={"zyte_api": ...}` requests. For example, someone may want to set screenshot options.

Thoughts?

> I also wonder, if users want automated mapping for specific requests, what would the meta look like?

Hm, `"zyte_api_transparent_mode": {"browserHtml": True}` makes sense, but not we have 2 way to do the same - 2nd is via `zyte_api_extra`. Or do yuou propose not to use `"zyte_api_extra"` name? This could work. But ``"zyte_api_transparent_mode"` name is very long; unlike function names, this is actually an issue, because it's easy to make a typo, and it's much harder to get auto-completion.

Reply:
OK, so what about this:
- `ZYTE_API_ENABLED` (default: `True`) toggles the plugin, but has no effect beyond that: requests still do not go through Zyte Data API unless request metadata demands so. This is mostly a way to allow easily disabling the plugin.
- Request metadata key `zyte_api` enables Zyte Data API for a request with the request URL and the specified parameters, merged with `ZYTE_API_DEFAULT_PARAMS`.
- There is no way to make all requests go through Zyte Data API with `ZYTE_API_DEFAULT_PARAMS`. We can consider adding it in the future if we see a valid use case. The workaround would be to set `"zyte_api": True` on all requests.
- Request metadata key `zyte_api_automap` enables Zyte Data API for a request with automated mapping, and can be used to pass additional parameters for that mode. Combining it with `zyte_api` causes an error. It is merged with `ZYTE_API_AUTOMAP_PARAMS`. 
- `ZYTE_API_TRANSPARENT_MODE` (default: `False`) makes it so that by default all requests are sent through Zyte Data API with automatic request parameter mapping enabled by default. Request metadata key `zyte_api_automap` can be used to provide extra parameters, or to disable automatic mapping for specific requests (and Zyte Data API usage, unless `zyte_api` is also used). Use of `zyte_api` on a request implies `"zyte_api_automap": False` for that request.

I wonder if we should deprecate and replace `zyte_api` and `ZYTE_API_DEFAULT_PARAMS` with `zyte_api_raw` and `ZYTE_API_RAW_PARAMS`. I think “raw” may help in fighting people’s expectations of automated parameter mapping.

We could also consider replacing `ZYTE_API_TRANSPARENT_MODE` with an enum-based `ZYTE_API_MODE` supporting values `manual` (default) and `transparent`, which we could later extend to support an additional value (e.g. `all`, or `automatic`) causing all requests to go through Zyte Data API with automated mapping disabled by default, i.e. using `ZYTE_API_DEFAULT_PARAMS`.

Reply:
>   ZYTE_API_ENABLED (default: True) toggles the plugin, but has no effect beyond that: requests still do not go through Zyte Data API unless request metadata demands so. This is mostly a way to allow easily disabling the plugin.

Makes sense.


> Request metadata key zyte_api enables Zyte Data API for a request with the request URL and the specified parameters, merged with ZYTE_API_DEFAULT_PARAMS.

Makes sense. 

> There is no way to make all requests go through Zyte Data API with ZYTE_API_DEFAULT_PARAMS. We can consider adding it in the future if we see a valid use case. The workaround would be to set "zyte_api": True on all requests.

+1; we can consider it later if there is a use case. It's also trivial for the users to write a middleware which does it.

> Request metadata key zyte_api_automap enables Zyte Data API for a request with automated mapping, and can be used to pass additional parameters for that mode. Combining it with zyte_api causes an error. It is merged with ZYTE_API_AUTOMAP_PARAMS.

Depending on the implementation, it may contradict the next point. But overall, it sounds good.

> ZYTE_API_TRANSPARENT_MODE (default: False) makes it so that by default all requests are sent through Zyte Data API with automatic request parameter mapping enabled by default. Request metadata key zyte_api_automap can be used to provide extra parameters, or to disable automatic mapping for specific requests (and Zyte Data API usage). Use of zyte_api on a request implies "zyte_api_automap": False for that request.

Makes sense.

> I wonder if we should deprecate and replace zyte_api and ZYTE_API_DEFAULT_PARAMS with zyte_api_raw and ZYTE_API_RAW_PARAMS. I think “raw” may help in fighting people’s expectations of automated parameter mapping.

That's reasonable, but more verbose. I think if we go this way, we need to think how can we simplify non-transparent requests further (browserHtml, data extraction).

Initially I had a design similar to scrapy-splash in mind:

* zyte_api meta key is the source of truth. It provides a raw API request (see https://github.com/scrapy-plugins/scrapy-zyte-api/issues/48), maybe without "url" though, not sure.
* Automatic mapping works by updating zyte_api meta key.
* Transparent mode works by updating zyte_api meta key.
* To support automatic mapping wihout transparent mode, and to simplify request building (no typos in string meta keys), we introduce a custom ZyteApiRequest.

I'm not sure it's the best though, given the issues we had with scrapy-splash implementation - it's quite hard to follow (partially because of mutability), and it lead to security issues. It also doesn't solve the problem of disabling individual requests when transparent mode is enabled, though it's solvable. Two separate meta keys may be fine. 

> We could also consider replacing ZYTE_API_TRANSPARENT_MODE with an enum-based ZYTE_API_MODE supporting values manual (default) and transparent, which we could later extend to support an additional value (e.g. all, or automatic) causing all requests to go through Zyte Data API with automated mapping disabled by default, i.e. using ZYTE_API_DEFAULT_PARAMS.

This sounds ok, thought I don't see big benefits.